presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayEqualOperator.java

```diff
@@ -47,15 +53,39 @@ private ArrayEqualOperator()
     @Override
     public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
     {
-        Type type = types.get("T");
-        type = typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(type.getTypeSignature()), ImmutableList.of());
+        Type elementType = types.get("T");
+        Type type = typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(elementType.getTypeSignature()), ImmutableList.of());
         TypeSignature typeSignature = type.getTypeSignature();
-        return operatorInfo(EQUAL, RETURN_TYPE, ImmutableList.of(typeSignature, typeSignature), METHOD_HANDLE.bindTo(type), false, ImmutableList.of(false, false));
+        MethodHandle equalsFunction = functionRegistry.resolveOperator(EQUAL, ImmutableList.of(elementType, elementType)).getMethodHandle();
+        MethodHandle methodHandle = methodHandle(ArrayEqualOperator.class, "equals", MethodHandle.class, Type.class, Slice.class, Slice.class);
+        MethodHandle method = methodHandle.bindTo(equalsFunction).bindTo(elementType);
+        return operatorInfo(EQUAL, RETURN_TYPE, ImmutableList.of(typeSignature, typeSignature), method, false, ImmutableList.of(false, false));
     }
 
-    public static boolean equals(Type type, Slice left, Slice right)
+    public static boolean equals(MethodHandle equalsFunction, Type type, Slice left, Slice right)
     {
-        //TODO: This could be quite slow, it should use parametric equals
-        return type.equalTo(createBlock(type, left), 0, createBlock(type, right), 0);
+        Block leftArray = readStructuralBlock(left);
+        Block rightArray = readStructuralBlock(right);
+        if (leftArray.getPositionCount() != rightArray.getPositionCount()) {
+            return false;
+        }
+        for (int i = 0; i < leftArray.getPositionCount(); i++) {
+            checkElementNotNull(leftArray.isNull(i), ARRAY_NULL_ELEMENT_MSG);
+            checkElementNotNull(rightArray.isNull(i), ARRAY_NULL_ELEMENT_MSG);
+            Object leftElement = castValue(type, leftArray, i);
```

Comment:
I'm concerned that this is going to hurt performance because it will box any primitives. Have you done any benchmarking for this?